controller/src/main/java/io/pravega/controller/store/stream/Int96CounterImpl.java

```diff
@@ -0,0 +1,140 @@
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.tracing.TagLogger;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+  * Int96CounterImpl implements getNextCounter using the Int96Counter interface.
+  */
+class Int96CounterImpl implements Int96Counter {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(Int96CounterImpl.class));
+
+    @VisibleForTesting
+    final BiFunction<CounterInfo, BiConsumer<Int96, Int96>, CompletableFuture<Void>> function;
+
+    private final Object lock;
+    @GuardedBy("lock")
+    private final AtomicInt96 limit;
+    @GuardedBy("lock")
+    private final AtomicInt96 counter;
+    private final SequentialProcessor sequentialProcessor;
+
+
+    public Int96CounterImpl(final BiFunction<CounterInfo, BiConsumer<Int96, Int96>, CompletableFuture<Void>> function,
+                            Executor executor) {
+        this.lock = new Object();
+        this.counter = new AtomicInt96();
+        this.limit = new AtomicInt96();
+        this.function = function;
+        this.sequentialProcessor = new SequentialProcessor(executor);
+    }
+
+    @Override
+    public CompletableFuture<Int96> getNextCounter(OperationContext context) {
+        CompletableFuture<Int96> future;
+        Int96 next;
+        synchronized (lock) {
+            next = counter.incrementAndGet();
+        }
+        if (next.compareTo(limit.get()) > 0) {
+            // ignore the counter value and after refreshing call getNextCounter
+            future = refreshRangeIfNeeded(context).thenCompose(x -> getNextCounter(context));
+        } else {
+            future = CompletableFuture.completedFuture(next);
+        }
+        return future;
+    }
+
+     @VisibleForTesting
+     void reset(Int96 previous, Int96 nextLimit) {
+        // Received new range, we should reset the counter and limit under the lock
+        // and then reset refreshfutureref to null
+        synchronized (lock) {
+            // Note: counter is set to previous range's highest value. Always get the
+            // next counter by calling counter.incrementAndGet otherwise there will
+            // be a collision with counter used by someone else.
+            counter.set(previous.getMsb(), previous.getLsb());
+            limit.set(nextLimit.getMsb(), nextLimit.getLsb());
+        }
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> refreshRangeIfNeeded(final OperationContext context) {
+        return sequentialProcessor.add(() -> {
+            // no ongoing refresh, check if refresh is still needed
+            if (counter.get().compareTo(limit.get()) >= 0) {
```

Comment:
same here. (Note you don't need a lock to guard an atomic if its ok that the reads/writes are interleaved, (I have not checked if that is true here), in which case just remove the annotation)

Reply:
Done